### PR TITLE
Zeroize values before deletion

### PR DESCRIPTION
**Pull request by ffflorian**

### CODE REVIEW

File: src/proteus/util/MemoryUtil.coffee [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+module.exports = do ->
+  zeroize_array: (object) ->
+    assert_is_zeroized = (array) ->
+      is_null = true
+      array.forEach (value) ->
+        if value > 0
+          is_null = false
+      if is_null isnt true
+        throw new ProteusError 'Array was not zeroized!'
```
Comment (herrmannplatz) [inline src/proteus/util/MemoryUtil.coffee]: what about using every? `array.every (value) -> value > 0`

Comment (ffflorian) [inline src/proteus/util/MemoryUtil.coffee]: Good idea!

File: src/proteus/derived/DerivedSecrets.coffee [hunk 2]
```diff
@@ -36,6 +37,8 @@ module.exports = class DerivedSecrets
     cipher_key = new Uint8Array okm.buffer.slice 0, 32
     mac_key = new Uint8Array okm.buffer.slice 32, 64
 
+    MemoryUtil.zeroize_buffer okm.buffer
```
Comment (bennyn) [inline src/proteus/derived/DerivedSecrets.coffee]: Please rename `okm` into `output_key_material`.

File: src/proteus/util/MemoryUtil.coffee [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          assert_is_zeroized val
+        else if typeof val is 'object'
+          memzero_all val
+
+    memzero_all object
+
+  zeroize_buffer: (buffer) ->
+    TypeUtil.assert_is_instance ArrayBuffer, buffer
+
+    bufferArray = new Uint8Array buffer
```
Comment (bennyn) [inline src/proteus/util/MemoryUtil.coffee]: Snake case here. 🐛 `bufferArray` -> `buffer_array`

File: src/proteus/util/MemoryUtil.coffee [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  assert_is_zeroized = (array) ->
+    array.every (value) ->
+      if value > 0
+        throw new ProteusError 'Array was not zeroized!'
+
+  zeroize_array: (object) ->
+    for key, val of object
+      if val instanceof Uint8Array
+        sodium.memzero val
+        assert_is_zeroized val
```
Comment (bennyn) [inline src/proteus/util/MemoryUtil.coffee]: The method `assert_is_zeroized` should be moved to test code.

File: src/proteus/util/MemoryUtil.coffee [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+TypeUtil = require '../util/TypeUtil'
+
+module.exports = do ->
+  assert_is_zeroized = (array) ->
+    array.every (value) ->
+      if value > 0
+        throw new ProteusError 'Array was not zeroized!'
+
+  zeroize_array: (object) ->
```
Comment (bennyn) [inline src/proteus/util/MemoryUtil.coffee]: Please write a unit test for `zeroize_array`.

**Pull request merged**